pkg/codegen/go/gen.go

```diff
@@ -2513,6 +2563,73 @@ func generatePackageContextMap(tool string, pkg *schema.Package, goInfo GoPackag
 		scanResource(r)
 	}
 
+	// compute set of names generated by a type
+	// handling any potential collisions via remapping along the way
+	scanType := func(t schema.Type) {
+		switch t := t.(type) {
+		case *schema.ObjectType:
```

Comment:
No other types allocate names except ObjectType? Enums don't?

Reply:
Unfortunately, like duplicate type names we also have some bespoke logic (and tests) in place for enums:

https://github.com/pulumi/pulumi/blob/4311a024cd5d120623936cb8ad8987149a2a5d22/pkg/codegen/go/gen.go#L1951

This change doesn't touch that, and is a targeted fix for the issues we're seeing in aws/gcp. Agree that we will likely need to cover more cases in the future, but will leave that as a future exercise as bug reports come in.